starbound/btreedb5.py

```diff
@@ -66,6 +66,41 @@ def get(self, key):
         # None of the keys in the leaf node matched.
         raise KeyError(binascii.hexlify(key))
 
+    def get_all_keys(self, start=None):
+        """
+        Returns a list of all valid keys starting at the given `start`
+        offset.  If `start` is `None`, we will start from the root of
+        the tree.
+        """
+        keys = []
+        s = self.stream
+        if not start:
+            start = HEADER_SIZE + self.block_size * self.root_block
+        s.seek(start)
+        block_type = s.read(2)
+        if block_type == LEAF:
+            reader = LeafReader(self)
+            num_keys = struct.unpack('>i', reader.read(4))[0]
+            for i in range(num_keys):
+                cur_key = reader.read(self.key_size)
+                keys.append(cur_key)
+                length = sbon.read_varint(reader)
+                reader.seek(length, 1)
+        elif block_type == INDEX:
+            (_, num_keys, first_child) = struct.unpack('>Bii', s.read(9))
+            children = [first_child]
+            for i in range(num_keys):
+                new_key = s.read(self.key_size)
+                next_child = struct.unpack('>i', s.read(4))[0]
+                children.append(next_child)
+            for child_loc in children:
+                keys.extend(self.get_all_keys(HEADER_SIZE + self.block_size * child_loc))
```

Comment:
Would you be opposed to making this method a generator instead? Basically, instead of creating a list of `keys`, just `yield cur_key` up in the `LEAF` case, and in the code right above this comment you can just re-`yield` the items*:

```python
for child_loc in children:
    for key in self.get_all_keys(HEADER_SIZE + self.block_size * child_loc):
        yield key
```

This would save a lot of allocations and copies of lists, and also speed up operations such as searching which don't require iteration of the entire list.

*) Since this code should be Python 2.x compatible you can't use the otherwise more idiomatic `yield from` :(